[FLINK-7153] Re-introduce preferred locations for scheduling

tillrohrmann opened the pull request:
## What is the purpose of the change

This PR makes the `TaskManagerLocation` accessible for asynchronous scheduling.

Due to changes for Flink 1.3 where we introduced asynchronous scheduling, it was not always guaranteed that the scheduler knew about the scheduling locations of producer tasks. Especially the eager scheduling mode was affected since the slot allocation happened concurrently.

In order to fix this problem, this PR adds a `TaskManagerLocationFuture` to each `Execution`. In eager scheduling mode, a slot will only be requested for a task if all its inputs have a slot assigned (e.g. their `TaskManagerLocationFuture` is completed). In lazy scheduling mode, we don't wait for the completion of all inputs, but take those inputs which are already known.

In order to distinguish whether we want to wait for all or take all available task manager locations, we add a `LocationPreferenceConstraint` which has the values `ALL` and `ANY`. `ALL` means that we have to wait for all inputs to have a location assigned, and `ANY` means that we take what's currently known.

In order to not deploy slots prematurely in eager mode, the slot assignment has been factored out into its own step. Before, one had to call `Execution#deployToSlot(SimpleSlot)` which assigned the given slot and started the deployment. Now, one has to call `Execution#tryAssignResource` before one can call `Execution#deploy`.

Moreover this PR fixes that the `FailoverRegions` are topologically sorted which is important for non queued scheduling.

FYI @StephanEwen 

## Brief change log

- Introduce `LocationPreferenceConstraint` to distinguish the waiting behaviour for the preferred locations
- Split slot assignment and deployment into two separate steps
- Moved preferred location calculation into the Execution to reduce code duplication between the `Scheduler` and the `SlotPool`
- Changed preferred location calculation to be blocking if `LocationPreferenceConstraint#ALL` and not all input locations are known

## Verifying this change

This change added tests and can be verified as follows:

- Added `ExecutionTest` to check the correct assigned slot release in case of cancellation and to check the correct preferred location calculation
- Added `ExecutionGraphDeploymentTest#testEagerSchedulingWaitsOnAllInputPreferredLocations` to check that eager scheduling waits for all inputs to be assigned before scheduling consumer tasks
- Moreover, the scheduler is being tested by existing tests such as `SchedulerSlotSharingTest`, `ScheduleWithCoLocationHintTest` and many IT cases for lazy scheduling (batch case)

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

summerleafs commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionGraph.java:
```diff
@@ -878,113 +880,70 @@ private void scheduleEager(SlotProvider slotProvider, final Time timeout) {
-			final ArrayList<CompletableFuture<SimpleSlot>> slotFutures = new ArrayList<>(getNumberOfExecutionJobVertices());
+		// allocate the slots (obtain all their futures
+		for (ExecutionJobVertex ejv : getVerticesTopologically()) {
+			// these calls are not blocking, they only return futures
```
Hi, allocate resources according to the order of topologically, is just to facilitate the optimization of 'allocate resource base on prefer input'? it may cause bad result if we allocate base on state.

tillrohrmann replied:
There is no specific reason why we iterate over the vertices in topological order. We could also choose a completely random order for eager scheduling because the scheduling order will be determined by the preferred location futures (which at the moment is based on inputs only). If we should switch to state location then it basically means that we schedule the individual tasks independently because the vertices don't depend on the input locations.

sihuazhou replied:
If we switch to state location then we can't allocate resources according to the order of topologically, because stateless vertices may share the same  SlotSharingGroup with stateful vertices, if stateless vertices allocated before the stateful vertices, the result can be bad.  An intuitive way to do this is to allocate resources to stateful vertices firstly.

tillrohrmann replied:
In that case, it depends a bit on how the scheduler values the state location preference. If it is implemented that it strictly schedules tasks to its previous state location, then it could happen that these tasks don't end up in the same slot as other tasks with which they shared a slot before.

tillrohrmann replied:
But be aware that the `allocateAndAssign` call is non-blocking and the actual order depends on the preferred locations futures.

StephanEwen replied:
Having a scheduling mechanism that tries to satisfy both state locality and input locality could be interesting. Input locality may extend to locality of input partitions (in Kafka for example) as well, which makes it even more complicated.

I think the current state of the heuristic is: previous location first (later: state locality first), if that leaves freedom, try to schedule based on inputs.

I can see an extended variant where we first collect all vertices with constraints, try to satisfy those. They may in turn add more constraints (or preferences) which should be satisfied next. Repeat until all are satisfied, or it is not possible to satisfy the preferences any more. But that is a pretty big change, that we should discus and design properly, not push it into a bug fix.

sihuazhou replied:
Aha, just by the way, scheduling according to the weight maybe not a bad choice. For both state and inputs, we can weigh them (maybe can weigh state according to it's size and weigh inputs according to it's throughput), then schedule according to the weight. 

This method can be easy to extend for other factors that we want to take account in scheduler.

summerleafs commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -1065,6 +1177,46 @@ private void sendUpdatePartitionInfoRpcCall(
+	@VisibleForTesting
+	public CompletableFuture<Collection<TaskManagerLocation>> calculatePreferredLocations(LocationPreferenceConstraint locationPreferenceConstraint) {
+		final Collection<CompletableFuture<TaskManagerLocation>> preferredLocationFutures = getVertex().getPreferredLocationsBasedOnInputs();
+		final CompletableFuture<Collection<TaskManagerLocation>> preferredLocationsFuture;
```
Hi Till,`getPreferredLocations()` is not invoked here because flink doesn't yet support reading the checkpoint data locally? I have create a issue for flink reading checkpoint locally [here](https://issues.apache.org/jira/browse/FLINK-7873?filter=-1), when it complete i wonder if we can invoke `getPreferedLocations()` instead of `getPreferredLocationsBasedOnInputs()`.

tillrohrmann replied:
Yes, that is exactly the idea.

sihuazhou commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/ExecutionVertex.java:
```diff
@@ -476,14 +482,13 @@ else if (numSources < parallelism) {
 		else {
-			Set<TaskManagerLocation> locations = new HashSet<>();
-			Set<TaskManagerLocation> inputLocations = new HashSet<>();
+			Set<CompletableFuture<TaskManagerLocation>> inputLocations = new HashSet<>(4);
```
I think if we weigh inputs in scheduler maybe this problem can be solved. In above example, imagine that, a is "broadcast" and it downstream‘s parallelism is 10, and b is "forward" , then every ExecutionEdge for a's weight is ```a-total-weight / 10``` and ExecutionEdge of b's weight is ```b-total-weight / 1```, we can evaluate total weight according to vertex's throughput, currently, all vertex's total weight are equal, so location of b's input will be picked.

tillrohrmann replied:
This change was not entirely intended and the previous code makes totally sense with your explanation. I'm actually in favour of reverting my changes to not change the semantics for the moment.

However, for the future, I'm wondering whether this kind of decision should be made by the `ExecutionVertex` or whether it shouldn't rather be the task of the `Scheduler` to make this kind of decision. 

For example, what if a task has multiple input gates and one of them with exactly one producer. Then it will only return the location of this single producer. Now if this TM has no more slots left, then we would basically randomly pick another slot even though there might be other TMs one which another producer for this task would run.

tillrohrmann replied:
For example in the case of the broadcast join with lazy scheduling, it could be the case that the broadcasting operator produces first and thus triggers the `scheduleOrUpdateConsumers` call on the `ExecutionGraph`. This will then trigger the scheduling of the join operator. At this time, there might only be the location of the broadcast operator known. However, since we only return the forward operator's location future which has not been completed, the lazy scheduling will schedule without location preference because the `LocationPreferenceConstraint` is `ANY`.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -126,9 +134,11 @@
 	private volatile ExecutionState state = CREATED;
 
-	private volatile SimpleSlot assignedResource;     // once assigned, never changes until the execution is archived
+	private final AtomicReference<SimpleSlot> assignedResource;
```
Good point. I'll change it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -220,14 +233,54 @@ public long getGlobalModVersion() {
+	 */
+	@VisibleForTesting
+	boolean tryAssignResource(final SimpleSlot slot) {
+		Preconditions.checkNotNull(slot);
```
Sure, will reformat it accordingly.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -1065,6 +1177,46 @@ private void sendUpdatePartitionInfoRpcCall(
+			case ALL:
+				preferredLocationsFuture = FutureUtils.combineAll(preferredLocationFutures);
+				break;
+			case ANY:
```
Yes that is the semantics of `ANY` at the moment. You basically do the scheduling decision based on what you currently know. In the worst case, you don't know any input location yet. I had to introduce this mode for lazy scheduling. In this case, however, you should always know at least one input location.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/executiongraph/Execution.java:
```diff
@@ -1065,6 +1177,46 @@ private void sendUpdatePartitionInfoRpcCall(
+				preferredLocationsFuture = FutureUtils.combineAll(preferredLocationFutures);
+				break;
+			case ANY:
+				final ArrayList<TaskManagerLocation> completedTaskManagerLocations = new ArrayList<>(1);
```
No, the intention is to return all currently known locations here. Usually there will only be one input because `ANY` is used by lazy scheduling. Initializing it with size 1 is a compromise between size and resizing costs. We could also initialize it with the number of inputs (since this is a small number) but in most cases not all inputs will be known for `ANY`.

tillrohrmann replied:
Maybe it's better to initialize the array with the number of returned futures to avoid resizing completely.

StephanEwen replied:
I would almost leave it at `1` or at the default `10`, so as to not have large unused arrays lying around.

tillrohrmann commented on flink-runtime/src/test/resources/log4j-test.properties:
```diff
@@ -16,7 +16,7 @@
 ################################################################################
 
-log4j.rootLogger=OFF, console
+log4j.rootLogger=INFO, console
```
Good catch. Will revert it.

tillrohrmann commented on flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/Scheduler.java:
```diff
@@ -133,32 +133,33 @@ public void shutdown() {
 				return CompletableFuture.completedFuture((SimpleSlot) ret);
-			}
-			else if (ret instanceof CompletableFuture) {
+			} else if (ret instanceof CompletableFuture) {
```
Beauty lies in the eye of the bee holder ;-) I'll revert it.

StephanEwen replied:
No worries, just teasing here ;-)

tillrohrmann commented:
Thanks a lot for your review @StephanEwen and @sihuazhou. I've addressed all your comments @StephanEwen. 

Next thing I'll do is to rebase onto the latest master and if Travis gives green light and you have no further objections, then I would like to merge it.

StephanEwen commented:
Thanks for the follow-up!

+1 to merge this
